src/client/sandbox/cookie/cookie-sync.js

```diff
@@ -0,0 +1,126 @@
+import nativeMethods from '../../sandbox/native-methods';
+import XhrSandbox from '../../sandbox/xhr';
+import settings from '../../settings';
+import { stringify as stringifyJSON } from '../../json';
+import { isIE9 } from '../../utils/browser';
+
+const SWITCH_BACK_TO_ASYNC_XHR_DELAY = 2000;
+
+export default class CookieSync {
+    constructor () {
+        this.useAsyncXhr = true;
+        this.queue       = [];
+        this.sendedQueue = null;
+        this.activeReq   = null;
+
+        // NOTE: When unloading, we should switch to synchronous XHR to be sure that we won't lose any cookies.
+        nativeMethods.windowAddEventListener.call(window, 'beforeunload', () => this._beforeUnloadHandler(), true);
+    }
+
+    static _createXMLHttpRequest (isAsync) {
+        var xhr = XhrSandbox.createNativeXHR();
+
+        xhr.open('POST', settings.get().cookieSyncUrl, isAsync);
+        xhr.setRequestHeader('Cache-Control', 'no-cache, no-store, must-revalidate');
+
+        return xhr;
+    }
+
+    static _has204StatusCode (request) {
+        try {
+            // NOTE: XMLHTTPRequest implementation in MSXML HTTP does not handle HTTP responses
+            // with status code 204 (No Content) properly; the `status' property has the value 1223.
+            return request.status === (isIE9 ? 1223 : 204);
+        }
+        catch (e) {
+            return false;
+        }
+    }
+
+    _beforeUnloadHandler () {
+        if (this.useAsyncXhr) {
+            this.useAsyncXhr = false;
+
+            // NOTE: When active request is aborted we will send message queue to the server
```

Comment:
Let's improve the comment.
We abort the current async request and resend message queue to a server using sync request.